### PR TITLE
Add dataset cache

### PR DESCRIPTION
**Pull request by davidotte**
### Change Description

*Try to be precise. You can additionally add comments to your PR, this might help the reviewer a lot.*

This PR adds a dataset cache to prevent repeated upload of datasets. There are also few changes in the corresponding [server PR](https://github.com/automl/tabpfn-server/pull/80). More concretely, this PR adds the following things:
- Cache system: The DatasetCacheManager keeps track of the cache in form of an OrderedDict, which contains a map of dataset hash to dataset uid for the last 50 uploaded datasets. To avoid specific edge cases (like change of user account or same test set for different train sets), the dataset is hashed together with the user's token and train_set_uid (in the case of the test set). For hashing I decided to use CityHash128 because in my tests it was way faster for large datasets than other common hashing algorithms. The cache is stored in a file in the .tabpfn folder.
- Caching in train set upload: When fit is called and the train set is in the cache, then no request is sent to the server at all.
- Caching in predict: If the test set is cached, then test_set_uid is sent instead of the actual test set. This part involves now a retry mechanism (implemented with the for loop) in case of any failures with cashing. This could for instance be needed if any datasets on server side were deleted by us or the cache file was manually changed. But this should almost never happen. Then, either cashed train_set_uid or test_set_uid can be incorrect, so both datasets are uploaded again.

Please check this PR carefully for the case that I missed any further edge cases :)

**If you used new dependencies: Did you add them to `requirements.txt`?**

yes

**Who did you ping on Mattermost to review your PR? Please ping that person again whenever you are ready for another review.**

@SamuelGabriel 

## Breaking changes

If you made any breaking changes, please update the version number.
Breaking changes are totally fine, we just need to make sure to keep the users informed and the server in sync.

**Does this PR break the API? If so, what is the corresponding server commit?**

**Does this PR break the user interface? If so, why?**

Yes, because now just the cached dataset uids might be sent to server.

---
*Please do not mark comments/conversations as resolved unless you are the assigned reviewer. This helps maintain clarity during the review process.*

### CODE REVIEW

File: tabpfn_client/client.py [hunk 1]
```diff
@@ -107,20 +186,33 @@ def upload_train_set(self, X, y) -> str:
...
-
-        X = common_utils.serialize_to_csv_formatted_bytes(X)
-        y = common_utils.serialize_to_csv_formatted_bytes(y)
+        # Save until prediction for retrying train set upload for the case that anything went wrong with cache.
+        self._cache_manager.temporary_save_train_set(X, y)
+        X_serialized = common_utils.serialize_to_csv_formatted_bytes(X)
+        y_serialized = common_utils.serialize_to_csv_formatted_bytes(y)
+
+        # Get hash for dataset. Include access token for the case that one user uses different accounts.
+        dataset_hash = str(
```
Comment (SamuelGabriel) [inline tabpfn_client/client.py:195]: I think the hashing could very well go inside the `DatasetUIDCache`, the function should just accept the X,y and extra strings to incorporate (e.g. acccess_token or train_uid)

File: tabpfn_client/client.py [hunk 2]
```diff
@@ -146,29 +238,66 @@ def predict(
             The predicted class labels.
         """
 
-        x_test = common_utils.serialize_to_csv_formatted_bytes(x_test)
+        x_test_serialized = common_utils.serialize_to_csv_formatted_bytes(x_test)
 
-        params = {"train_set_uid": train_set_uid, "task": task}
+        # Get hash for dataset. Include train_set_uid for the case that the same test set was previously used
+        # with different train set. Include access token for the case that a user uses different accounts.
+        dataset_hash = str(
```
Comment (SamuelGabriel) [inline tabpfn_client/client.py:245]: this hashing could also go to the cache manager

File: tabpfn_client/client.py [hunk 3]
```diff
@@ -71,6 +149,7 @@ def __init__(self):
         )
 
         self._access_token = None
+        self._cache_manager = DatasetCacheManager()
```
Comment (SamuelGabriel) [inline tabpfn_client/client.py:152]:
```suggestion
        self._dataset_uid_cache_manager = DatasetCacheManager()
```

File: tabpfn_client/client.py [hunk 4]
```diff
@@ -146,29 +238,66 @@ def predict(
...
+                    train_set_uid = self.upload_train_set(X_train, y_train)
+                    self._cache_manager.add_dataset(old_train_set_hash, train_set_uid)
+                    params["train_set_uid"] = train_set_uid
+                    cached_test_set_uid = None
+                else:
+                    raise  # Re-raise the exception if it's not recoverable
 
-        self._validate_response(response, "predict")
+        else:
+            raise RuntimeError("Failed to get prediction after retries.")
```
Comment (SamuelGabriel) [inline tabpfn_client/client.py:291]:
```suggestion
            raise RuntimeError(f"Failed to get prediction after {max_attempts} retries.")
```

**Review by SamuelGabriel — CHANGES_REQUESTED** [review]
This looks generally nice. There are two weak points that would need to be addressed though, I believe. There seems to be a slight bug, when using multiple estimator objects, and the naming is oftentimes hard to get, as it confuses "Dataset" and "DatasetUID".

File: tabpfn_client/client.py [hunk 5]
```diff
@@ -32,6 +36,80 @@ def filter(self, record):
         return True
 
 
+class DatasetCacheManager:
```
[the remark this answers is not in the document]
Comment (davidotte) [reply tabpfn_client/client.py]: Thanks for the review, good catch with the bug! I now moved the last train set part to the estimators, I feel like this is more straightforward with our current implementation than including it in the InferenceClient. I also did some renaming to avoid confusion. :)

File: tabpfn_client/client.py [hunk 6]
```diff
@@ -32,6 +36,80 @@ def filter(self, record):
...
+        os.makedirs(os.path.dirname(self.file_path), exist_ok=True)
+        with open(self.file_path, "w") as file:
+            json.dump(self.cache, file)
+
+    def delete_dataset_by_uid(self, dataset_uid: str) -> Optional[str]:
+        """
+        Deletes an entry from the cache based on the dataset UID.
+        """
+        hash_to_delete = None
+        for hash, uid in self.cache.items():
```
[the remark this answers is not in the document]
Comment (davidotte) [reply tabpfn_client/client.py:101]: I think we implicitly have this rule. When adding a dataset_uid to the dictionary, the old value is automatically overwritten.

File: tabpfn_client/client.py (hunk 1 above)
Comment (davidotte) [reply tabpfn_client/client.py]: I also thought about this, I was a bit hesitant to do that because we hash very different kind of stuff in fit and predict. But I agree with you. I moved the hashing to the DatesetUIDCacheManager and simply combine all arguments for the hash. Still not 100% sure how intuitive my new implementation is

Comment (SamuelGabriel) [reply tabpfn_client/client.py]: 👍

**Review by SamuelGabriel — APPROVED** [review]
Thanks for all the fixes. This PR looks very good now.
